lib/oauth2/request.ex

```diff
@@ -1,39 +1,79 @@
 defmodule OAuth2.Request do
   @moduledoc false
 
-  use HTTPoison.Base
-
   import OAuth2.Util
 
-  alias OAuth2.Error
-  alias OAuth2.Response
+  alias OAuth2.{Client, Error, Response}
+
+  @type body :: any
 
-  def request(method, url, body \\ "", headers \\ [], opts \\ []) do
+  @doc """
+  Makes a request of given type to the given URL using the `OAuth2.AccessToken`.
+  """
+  @spec request(atom, Client.t, binary, body, Client.headers, Keyword.t) :: {:ok, Response.t} | {:error, Error.t}
+  def request(method, %Client{} = client, url, body, headers, opts) do
+    url = client |> process_url(url) |> process_params(opts[:params])
+    headers = req_headers(client, headers)
     content_type = content_type(headers)
-    body = process_request_body(body, content_type)
+    body = encode_request_body(body, content_type)
     headers = process_request_headers(headers, content_type)
 
-    case super(method, url, body, headers, opts) do
-      {:ok, %HTTPoison.Response{status_code: status, headers: headers, body: body}} ->
+    case :hackney.request(method, url, headers, body, opts ++ [with_body: true]) do
```

Comment:
Adding the `:with_body` option to the end, might break if someone passes `[with_body: false]` in the options. Prepending it makes sure, it's always there.